EasyCommands/BlockHandlers/TerminalBlockHandlers.cs

```diff
@@ -19,38 +19,41 @@
-                KeyValuePair("Boolean", PropertyConverter((b, p) => b.GetValueBool(p), (b, p, v) => b.SetValueBool(p, CastBoolean(v)))),
-                KeyValuePair( "Single", PropertyConverter((b, p) => b.GetValueFloat(p), (b, p, v) => b.SetValueFloat(p, CastNumber(v)))),
-                KeyValuePair( "Int64", PropertyConverter((b, p) => (float)b.GetValue<long>(p), (b, p, v) => b.SetValue(p, (long)CastNumber(v)))),
-                KeyValuePair("Color", PropertyConverter((b, p) => b.GetValueColor(p), (b, p, v) => b.SetValueColor(p, CastColor(v))))
+        public class TerminalPropertyHandler<T> : SimplePropertyHandler<T> where T : class, IMyTerminalBlock {
+            public TerminalPropertyHandler(String propertyId, Primitive delta) : this(new PropertySupplier(propertyId), delta) { }
+            public TerminalPropertyHandler(PropertySupplier propertySupplier, Primitive delta) : base(
+                (b, p) => { var property = GetTerminalProperty(b, propertySupplier); return ResolvePrimitive(TerminalPropertyConversions[property.TypeName].GetValue(property, b));  },
+                (b, p, v) => { var property = GetTerminalProperty(b, propertySupplier); TerminalPropertyConversions[property.TypeName].SetValue(property, b, v); },
+                delta) { }
+
+            static Dictionary<String, TerminalPropertyConverter> TerminalPropertyConversions = NewDictionary(
+                KeyValuePair("StringBuilder", PropertyConverter((p, b) => p.As<StringBuilder>().GetValue(b).ToString(), (p, b, v) => p.As<StringBuilder>().SetValue(b, new StringBuilder(CastString(v))))),
+                KeyValuePair("Boolean", PropertyConverter((p, b) => p.AsBool().GetValue(b), (p, b, v) => p.AsBool().SetValue(b, CastBoolean(v)))),
+                KeyValuePair("Single", PropertyConverter((p, b) => p.AsFloat().GetValue(b), (p, b, v) => p.AsFloat().SetValue(b, CastNumber(v)))),
+                KeyValuePair("Int64", PropertyConverter((p, b) => (float)p.As<long>().GetValue(b), (p, b, v) => p.As<long>().SetValue(b, (long)CastNumber(v)))),
+                KeyValuePair("Color", PropertyConverter((p, b) => p.AsColor().GetValue(b), (p, b, v) => p.AsColor().SetValue(b, CastColor(v))))
             );
 
-            static TerminalPropertyConverter GetPropertyConverter(T block, PropertySupplier propertySupplier) {
-                var property = block.GetProperty(propertySupplier.propertyType);
-                if (property == null) throw new Exception(block.BlockDefinition.SubtypeName + " does not have property: " + (propertySupplier.propertyWord ?? propertySupplier.propertyType));
-                return TerminalPropertyValueConversion[property.TypeName];
-            }
+            static ITerminalProperty GetTerminalProperty(T block, PropertySupplier propertySupplier) =>
+                PROGRAM.propertyCache.GetOrCreate(typeof(T), propertySupplier.propertyType, s =>{
```

Comment:
Are we sure that the properties for a given block are always the same when queried by class type?  For example, the Projector and Console Block have the same block type.  Are we sure the properties will be the same?

I'm a little concerned about use cases where we would query a sub block as a "IMyTerminalBlock" block, and then cache a property by type for IMyTerminalBlock incorrectly.  

Example:

```
set "My Beacon" terminal "Range" property to 200
set "My Gatling Turret" terminal "Range" property to 200
```

In both cases we are caching the "Range" property for the IMyTerminalBlock (I believe) but the property for the individual blocks is actually different.

Reply:
Terminal properties and actions are only defined by block type. So that is not the problem.
But as you pointed out he code above might do the wrong thing. Replacing `typeof(T)` by `b.GetType()` will get the `Type` instance for the actual block type.

Reply:
Ya the b.GetType() should be more accurate.

Let's add a few tests to make sure that caching is working as so we don't break it in the future somehow.